draft-yasskin-http-origin-signed-responses.md

```diff
@@ -573,6 +573,110 @@ Signature: sig3;
+that the `Accept-Signature` header field indicates the client supports. However,
+if the `Accept-Signature` value violates a requirement in this section, the
+server MUST NOT send any `Signature` headers.
+
+The `Accept-Signature` header field is a Structured Header as defined by
+{{!I-D.ietf-httpbis-header-structure}}. Its value MUST be a list (Section 4.8 of
+{{!I-D.ietf-httpbis-header-structure}}) of parameterised labels (Section 4.4 of
+{{!I-D.ietf-httpbis-header-structure}}). The order of labels in the
+`Accept-Signature` list is not significant. Labels, ignoring any initial "-"
+character, MUST NOT be duplicated.
+
+Each label in the `Accept-Signature` header field's value indicates that a
+feature of the `Signature` header field ({{signature-header}}) is supported. If
+the label begins with a "-" character, it instead indicates that the feature
+named by the rest of the label is not supported. Unknown labels and parameters
+MUST be ignored because new labels and new parameters on existing labels may be
+defined by future specifications.
+
+### Integrity labels ### {#accept-signature-integrity}
+
+Labels starting with "digest/" indicate that the client supports the `Digest`
+header field ({{!RFC3230}}) with the digest-algorithm from the
+<https://www.iana.org/assignments/http-dig-alg/http-dig-alg.xhtml> registry
+named in lower-case by the rest of the label. For example, "digest/sha-512"
+indicates support for the SHA-512 digest algorithm, and "-digest/sha-256"
+indicates non-support for the SHA-256 digest algorithm.
+
+Labels starting with "mi/" indicate that the client supports the `MI` header
+field ({{!I-D.thomson-http-mice}}) with the parameter from the HTTP MI Parameter
+Registry registry named in lower-case by the rest of the label. For example,
+"mi/mi-blake2" indicates support for Merkle integrity with the
+as-yet-unspecified mi-blake2 parameter, and "-digest/mi-sha256" indicates
+non-support for Merkle integrity with the mi-sha256 content encoding.
+
+If the `Accept-Signature` header field is present, servers SHOULD assume support
+for "digest/sha-256" and "mi/mi-sha256" unless the header field states
+otherwise.
+
+### Key type labels ### {#accept-signature-key-types}
+
+Labels starting with "rsa/" indicate that the client supports certificates
+holding RSA public keys with a number of bits indicated by the digits after the
+"/".
+
+Labels starting with "ecdsa/" indicate that the client supports certificates
+holding ECDSA public keys on the curve named in lower-case by the rest of the
+label.
+
+If the `Accept-Signature` header field is present, servers SHOULD assume support
+for "rsa/2048", "ecdsa/secp256r1", and "ecdsa/secp384r1" unless the header field
+states otherwise.
+
+### Key value labels ### {#accept-signature-key-values}
+
+The "ed25519key" label has parameters indicating the public keys that will be
+used to validate the returned signature. Each parameter's name is re-interpreted
```

Comment:
Would it be possible for the client (Chromium) to request ed25519key signature response, while not specifying the ed25519 public key (or its prefix)?
I'm not sure if the UA knows which key its going to use at the request time (similar to the prefetch argument above).

Reply:
Right now, no, although we could specify a way to say that. For example, `ed25519key; *`. I *think* that for the https://github.com/mikewest/signature-based-sri use case, the `integrity=ed25519-...` attribute is always available next to the `src`, so the browser could include the public key.
@otherdaniel

Reply:
According to Mike's proposal, this would always be the case.

I can imagine use cases where this wouldn't be the case. Particularly if [Origin Policy](https://wicg.github.io/origin-policy/) happens, it'd be very convenient if the origin could declare its supported keys, without requiring them to be duplicated within each page. That would then break this.

A related, but somewhat incomplete thought: Mike has been pushing around the idea that maybe each resource should carry its own integrity assertions (that is, the resource declares it's own hash, or maybe signature, etc.), which would make it somewhat self-contained. Then, much of the work could happen as each resource is downloaded, independently of where/how it's being used. That's a somewhat incomplete plan, but IMHO rather attractive since it offers a path to bring the cost of (additional) security down to near-zero, and predictably so. Much of what we discuss here isn't directly contradicting this, but does go in the opposite direction, in that a resource could only be integrity-checked if requested in specific ways.

Reply:
Yep, I wouldn't want to get in the way of putting keys in the Origin Policy. Servers can definitely send the `Signature` even if the browser didn't explicitly request it, which fits the second half. If that becomes policy, we can drop the `ed25519key; ...` stuff entirely.